src/goto-symex/symex_goto.cpp

```diff
@@ -353,138 +353,155 @@ void goto_symext::merge_value_sets(
   dest.value_set.make_union(src.value_set);
 }
 
+/// Applies f to `(k, ssa, i, j)` if the first map maps k to (ssa, i) and
+/// the second to (ssa', j). If the first map has an entry for k but not the
+/// second one then j is 0, and when the first map has no entry for k then i = 0
+static void for_each2(
+  std::map<irep_idt, std::pair<ssa_exprt, unsigned>> first_map,
+  std::map<irep_idt, std::pair<ssa_exprt, unsigned>> second_map,
+  std::function<void(const irep_idt&, const ssa_exprt&, unsigned, unsigned)> f)
+{
+  auto second_it = second_map.begin();
+  for(const auto first_pair : first_map)
+  {
+    while(second_it != second_map.end() && second_it->first < first_pair.first)
+    {
+      f(second_it->first, second_it->second.first, 0, second_it->second.second);
+      ++second_it;
+    }
+    const ssa_exprt &ssa = first_pair.second.first;
+    const unsigned count = first_pair.second.second;
+    if(second_it != second_map.end() && second_it->first == first_pair.first)
+    {
+      f(first_pair.first, ssa, count, second_it->second.second);
+      ++second_it;
+    }
+    else
+      f(first_pair.first, ssa, count, 0);
+  }
+  while(second_it != second_map.end())
+  {
+    f(second_it->first, second_it->second.first, 0, second_it->second.second);
+    ++second_it;
+  }
+}
+
 void goto_symext::phi_function(
   const statet::goto_statet &goto_state,
   statet &dest_state)
 {
-  auto ssa_of_current_name =
-    [&](const std::pair<irep_idt, std::pair<ssa_exprt, unsigned>> &pair) {
-      return pair.second.first;
-    };
-
-  auto dest_state_names_range =
-    make_range(dest_state.level2.current_names)
-      .filter(
-        [&](const std::pair<irep_idt, std::pair<ssa_exprt, unsigned>> &pair) {
-          // We ignore the identifiers that are already in goto_state names
-          return goto_state.level2_current_names.count(pair.first) == 0;
-        })
-      .map<const ssa_exprt>(ssa_of_current_name);
-
-  // go over all variables to see what changed
-  auto all_current_names_range = make_range(goto_state.level2_current_names)
-                                   .map<const ssa_exprt>(ssa_of_current_name)
-                                   .concat(dest_state_names_range);
-
   guardt diff_guard;
-  if(!all_current_names_range.empty())
-  {
-    diff_guard=goto_state.guard;
+  if(goto_state.level2_current_names.empty() &&
+     dest_state.level2.current_names.empty())
+    return;
 
-    // this gets the diff between the guards
-    diff_guard-=dest_state.guard;
-  }
+  diff_guard=goto_state.guard;
+  // this gets the diff between the guards
+  diff_guard-=dest_state.guard;
+
+  for_each2(
+    goto_state.level2_current_names,
+    dest_state.level2.current_names,
+    [&](const irep_idt &l1_identifier,
+        const ssa_exprt &ssa,
+        unsigned goto_count,
+        unsigned dest_count)
+    {
```

Comment:
As a first refactoring, and without making any other changes, can we please move this loop body in a function of its own? This will make all other changes a lot easier to read.